### PR TITLE
sequence/proteome/:species GET endpoint for whole proteome download

### PR DESCRIPTION
**Pull request by vsitnik**
### Requirements

- Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion;
- Review the [contributing guidelines](https://github.com/Ensembl/ensembl/blob/release/90/CONTRIBUTING.md#why-could-my-pull-request-be-rejected) for this repository; remember in particular:
    - do not modify code without testing for regression
    - provide simple unit tests to test the changes
    - the PR must not fail unit testing
    - if you're adding/updating documentation of an endpoint, make sure you add/update the necessary parameters to the (template) configuration files in the ensembl-rest_private repo

### Description

New endpoint allows downloading all protein sequences for the specified species.
Only species having 'true' meta.proteome_download_allowed in the core databases will be affected.
For others this feature will be forbidden.

### Use case

Will be use by uniprot to download protein fastas from vectorbase.org.
`wget --header='Content-type:text/x-fasta' 'http://127.0.0.1:34274/sequence/proteome/Anopheles atroparvus?' -O - | gzip - > Aatr.prot.fasta.gz`

### Benefits

The endpoint allows to download all 'canonical' protein sequences for Anopheles atroparvus in 2 minutes 25 seconds instead of approximately 3 hours when using current approach.

### Possible Drawbacks

Still slow. Won't be appropriate for a large genomes, probably.
Thus, setting meta.proteome_download_allowed should be done with cautious.
seq_regions should have proper 'coding_cnt' and 'toplevel' attributes set. 

### Testing

`t/sequences.t` updated to test the new endpoint behaviour.
No regression was seen for the affected features. 

VectorBase prod db anopheles_atroparvus_core_1810_93_3 was used fot the performance testing. 

### Changelog

It's a new endpoint, which allows whole proteome fasta downloads.

### CODE REVIEW

File: lib/EnsEMBL/REST/Model/Registry.pm [hunk 1]
```diff
@@ -467,6 +481,7 @@ sub _build_species_info {
       assembly => $assembly_lookup{$species},
       strain => $strain_lookup{$species},
       strain_collection => $strain_collection_lookup{$species},
+      proteome_download_allowed => $proteome_download_allowed{$species},
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply lib/EnsEMBL/REST/Model/Registry.pm]: The problem is that there was (and still is) regression on the clean master. So I'll wait till you fix it and resubmit then.

[the remark this answers is not in the document]
Comment (vsitnik) [reply lib/EnsEMBL/REST/Model/Registry.pm]: fixed. no regression

File: t/sequences.t [hunk 2]
```diff
@@ -46,6 +46,7 @@ while(my $seq = $io->next_seq()) {
 } 
 
 my $dba = Bio::EnsEMBL::Test::MultiTestDB->new();
+my $dba2 = Bio::EnsEMBL::Test::MultiTestDB->new('anopheles_atroparvus');
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply t/sequences.t]: don't get it. test is failing without this line.

File: t/sequences.t [hunk 3]
```diff
@@ -411,6 +412,35 @@ FASTA
...
+  my $url = "/sequence/proteome/$species?";
+  action_check_code($url, 400, 'Return code for species with no meta.proteome_download_allowed');
+  action_raw_bad_regex($url, qr/Proteome download is not allowed for the species/, 'No proteome downloads if not allowed');
+}
+{
+  my $species = 'anopheles_atroparvus';
+  my $url = "/sequence/proteome/$species?";
+  my $fasta = fasta_GET($url, 'Getting whole proteome as a single fasta');
+  my $expected = <<'FASTA';
+>AATE007892-PA
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply t/sequences.t]: fixed

**Comment by nerdstrike** [comment]
@vsitnik As it stands, this is not a typical kind of endpoint that we support in Ensembl REST, so I can't recommend that we include this in our regular deployment. It monopolises resources for a long time, to the detriment of potentially thousands of other users. We've seen serious failures of service when long-running endpoints are in the regular pool. I am also uncomfortable with advertising functionality that is not available for the model organisms.

I see a few possible routes:

1) Include your endpoint, but add no documentation. Vectorbase can provide bulk protein downloads, nobody knows you can do it to other species. @lairdm tells me that you can have undocumented endpoints without breaking the server, so this is a practical solution on the proviso that it receives copious documentation to the effect. We may also want a server config switch that disables the endpoint to prevent accidental discovery and confusion. The EQTL controller exhibits such an on/off switch I believe.

2) Include your model changes only, then you create a controller in the same namespace but in another repository. As Vectorbase are running a distinct service, you would then install the repo alongside your REST installation, and that controller would be detected by Catalyst if it is in the PERL5LIB. Then the server acquires the extra endpoint without impacting regular Ensembl deployments. The downside is that this is a hacky behaviour and introduces extra installation detail for Vectorbase. The platform is quite stable at this stage, so you wouldn't have to keep fixing it, at least until 3)

3) Make do with FTP links until REST 2.0 comes to fruition, then develop data-hungry endpoints as required with whatever custom backend suits the data. It's still at the concept stage and not likely to materialise in a hurry.

**Comment by magaliruffier** [comment]
As there has been no activity in the last month, I will close this PR.
Please address all the comments before submitting a new request.

**Pull request closed without merging**